### ISSUE TITLE
Rewrite Escoria framework

### ISSUE THREAD

**Issue by StraToN** [godotengine/escoria#44]
We've come to a point where maintaining Escoria framework gets harder and harder. Escoria was written early in the history of Godot and has not evolved at the same pace. 

Godot Engine 3.0 stable release is very close, and we are facing many problems and bugs which are hard to solve (#43 https://github.com/flossmanualsfr/escoria/issues/12 #41 #35 ...)

To be able to clean the code from its useless parts and fix the most problematic bugs, I propose a complete rewrite of Escoria.

## Pros
- We will be able to provide a cleaner code thanks to Godot 3.0 evolutions, in an easier way than simply mutate the current code. Would fix #10 
- Reorganize elements so users know what included scripts/scenes they may need to modify. As much as possible, users should not modify base scripts and scenes.
- I'd personally like to "open" the global_vm (currently accessed through .esc files) to other ways, such as graph tools (for dialogs, game logic, items use, etc). 
- Keep .esc files as well. They're also a good choice for game scripting. Simply, adding more possibilities to achieve game scripting is, IMHO, a good thing for many users, especially if all solutions are compatible one with the other.
- We can integrate Escoria closer to the editor by creating a plugin (with custom nodes) and editor tools directly integrated in the AssetLib. Would fix #5 

## Cons
- As a new version, we do not need it to be fully compatible with previous versions. Thus projects based on current Escoria will have some hard time moving to it if they want to.

Share your ideas here.

# TODOs
- .esc files edition in Godot Script Editor (https://github.com/godotengine/godot/issues/1390)
- Define 1 unique script to manage items and NPC (name it "interactive" for example) and allow:
  - Wandering around by .esc script and/or Position2D/3D definition in the editor
  - ?
- Custom nodes:
  - RoomExit: define an Area2D>CollisionPolygon2D with a target scene.
  - ?
- Manage different types of terrains
  - 2D:
    - Background image
    - TileMap
  - 3D

# Target workflow:
![escoria1](https://user-images.githubusercontent.com/1289207/35041347-b62f03e4-fb84-11e7-84c9-e59c6061b0c6.png)

**Comment by fleskesvor** [comment]
I think this is a good idea, even if it's going to be a massive job. Are you aiming to keep the framework in a functional state during the process, and do you have ideas on how to best break the job up into actionable tasks?

> Reorganize elements so users know what included scripts/scenes they may need to modify. As much as possible, users should not modify base scripts and scenes.

I'm very much in favor of this. I think users are much more likely to have a positive development experience if they can make a game without also making a huge mess of their project files, and this can be enforced by improving the file structure. I think the biggest problem currently is that users have to make changes in the dialogue scenes under globals/ if they want to add extra dialoges, but there are also scripts under game/ and ui/, which I don't think a user would normally have any reason to edit. I'm also not sure if game/ really describes what its purpose is. It mostly deals with inventory items, so why aren't those scenes in ui/?

The demo scene could also do with some re-organizing to demonstrate re-use of scenes in a much better way by grouping different kinds of scenes (rooms, characters, items, etc.) under different paths.

> We can integrate Escoria closer to the editor by creating a plugin (with custom nodes) and editor tools directly integrated in the AssetLib. Would fix #5

Do you mean like an optional plugin on top of Escoria, that would be somewhat similar to tools like AGS and Visionaire? That would probably be great for attracting developers with little programming experience to Escoria. 

> As a new version, we do not need it to be fully compatible with previous versions. Thus projects based on current Escoria will have some hard time moving to it if they want to.

I agree. For instance, the new background area script is more flexible than the old one, so that should be the right way to do backgrounds. I think there are few enough Escoria projects in development that if anyone is having a hard time moving, we can probably help them all with the process.

**Comment by StraToN** [comment]
>I think this is a good idea, even if it's going to be a massive job. Are you aiming to keep the framework in a functional state during the process, and do you have ideas on how to best break the job up into actionable tasks?

The job is massive indeed.
I think we can keep the framework functional during the process, however it may provide less features than it currently does. For example, dialogs will probably not be available until many other elements have been rewritten. But other parts may be affected similarly.

>The demo scene could also do with some re-organizing to demonstrate re-use of scenes in a much better way by grouping different kinds of scenes (rooms, characters, items, etc.) under different paths.

Sure! The demo scene alone is insufficient, we need more of them, as we discussed this before. I got in touch with an artist (Livio Fania - http://liviofania.com/ ) who's interested in making new backgrounds, items and actors for the demo scene. He also made some propositions for a logo, but I am keeping this warm for the moment :) . He also introduced me with Fungus, a FLOSS toolkit for Unity that aims at pretty much the same thing as Escoria (https://fungusgames.squarespace.com/).



>>We can integrate Escoria closer to the editor by creating a plugin (with custom nodes) and editor tools directly integrated in the AssetLib. Would fix #5
>
>Do you mean like an optional plugin on top of Escoria, that would be somewhat similar to tools like AGS and Visionaire? That would probably be great for attracting developers with little programming experience to Escoria.

I'd like to get closer to these examples, yes, but I think Godot already does a lot of stuff by itself without us having to remake it (Navigation mesh, Sprite Animations...). So we may only need to add what's missing for our very purpose, here are some examples:
- dialog system, 
- inventory system, 
- automatization of links between scenes (like some sort of "Exit node" linking an exit area of a scene to another), 
- game designing tools (for example, graph based) for dialogs, story checkpoints and puzzles design.

All these ideas require the current VM to be reorganized and to provide an API for accessing it, so that you can control the game by all the ways you want: .esc scripts, GDScript (with direct calls to the VM) or any other system (anyone could create a plugin for the required purpose).

In this design, the VM is central. The VM manages inputs: 
- either player inputs (mouse clicks and such) 
- or script commands. 
And makes the game reacts to inputs. With this design, making the main actor walk towards a direction is achieved by 2 ways:
- either the player clicked at a position
- or an .esc script called a specific command like `walk_to player_id position.x position.y`
The result of both examples above would be identical.




>>As a new version, we do not need it to be fully compatible with previous versions. Thus projects based on current Escoria will have some hard time moving to it if they want to.
>
>I agree. For instance, the new background area script is more flexible than the old one, so that should be the right way to do backgrounds. I think there are few enough Escoria projects in development that if anyone is having a hard time moving, we can probably help them all with the process.

True. Even if it seems a good idea, I think it is wrong to correlate input catches into the same node that displays the background. Displaying the background is TextureRect's or Sprite's job. Catching inputs and transmit them to the VM is background_Area2D node's job. As such, since every scene is supposed to have a background_Area2D node, we probably shouldn't force users to create it themselves if we can do it for him.


Another point I'd like to think about is the ability to make Escoria compatible with both 2D and 3D.

**Comment by fleskesvor** [comment]
> Sure! The demo scene alone is insufficient, we need more of them, as we discussed this before. I got in touch with an artist (Livio Fania - http://liviofania.com/ ) who's interested in making new backgrounds, items and actors for the demo scene. He also made some propositions for a logo, but I am keeping this warm for the moment :) . He also introduced me with Fungus, a FLOSS toolkit for Unity that aims at pretty much the same thing as Escoria (https://fungusgames.squarespace.com/).

His art looks very good. It would be great to have a few demo scenes done in that style.

> All these ideas require the current VM to be reorganized and to provide an API for accessing it, so that you can control the game by all the ways you want: .esc scripts, GDScript (with direct calls to the VM) or any other system (anyone could create a plugin for the required purpose).

Ok, I think I'm starting to get a feel for how that can be organised in a more modular way:

* nodes which need to capture input (background, items, etc.) still connect to input events to a function as before, but this is to be considered local to the node
* events are passed to a singleton for processing, so that this code isn't repeated in every node, and  returns all relevant information about the event (eg. left-click, right-click, pressed, released, coordinates, etc.). This should make it easier to do "advanced" input handling, like drag events (item usage on mobile devices) and double-click (fast walk) since a singleton should allow us to keep track of state across events on different nodes.
* if the input was captured by a node, the now returned pre-processed event is passed on to a different function in the same node
* this node is considered exposed, or the input API of the node, and can also be called programmatically from the current `global_vm.gd`, which I agree should be renamed or, maybe better, be broken down into multiple APIs per feature, since it currently sort of duplicates the full API of `vm_level.gd`.

Of course, we need to use naming conventions, like an underscore prefix, and comments to signal which functions are to be considered local to a script and which are part of an API.

Also, these are just some quick thoughts on how it might be implemented, and it might not be the best way at all.

> automatization of links between scenes (like some sort of "Exit node" linking an exit area of a scene to another)

This might be useful for games using .esc scripts as well, and could be implemented as a special kind of interactive elements which only catches events when another interactive element (eg. item or player) is moved onto it. Like items, an exit might have use for sprites and animations, eg. to open and close a door, though it might just be a shape at the edge of a screen.

I agree with your other thoughts.

**Comment by fleskesvor** [comment]
I'm not sure how the capture-process-propagate dynamic would work on practice though, since you'd have to register multiple events before propagating it to the API function. Only propagating on release events would solve some of it, since that would let you aggregate multiple events, but what about double-clicks? And with drag events, you're not just interested in the start and end positions, since you'd want intermediate positions for moving a sprite across the screen during the event. I still think it's a good idea to use custom event types (defined as constants in the input singleton) for the API functions, rather than raw event types though. I think I might experiment a bit with this to see what's feasible.

**Comment by StraToN** [comment]
Here is a graph that shows the principles of my description. Let's use it as a basis.
![escoria1](https://user-images.githubusercontent.com/1289207/35041347-b62f03e4-fb84-11e7-84c9-e59c6061b0c6.png)
File for edition on http://draw.io 
[escoria.zip](https://github.com/godotengine/escoria/files/1638807/escoria.zip)

**Comment by fleskesvor** [comment]
I think that looks like a good basis, and it should probably not be too difficult to achieve.

In that model, is it the responsibility of the core to keep track of the selected input, eg. "look", so that an item/item_background when clicked mostly just needs to pass that event on to the core and then the core checks to see what the currently selected action is, if any?

Another thing this model reminded me of is that this would be a good opportunity to investigate the possibility of erasing some (all?) of the distinctions between items and and players and instead have either the core keep track of which item currently acts as the player or have that be a property of the item. The reason why this would be useful is because it would make it easier to do things like have an NPC walk to a specific point or even have multiple playable characters that you can switch between.

**Comment by StraToN** [comment]
> In that model, is it the responsibility of the core to keep track of the selected input, eg. "look", so that an item/item_background when clicked mostly just needs to pass that event on to the core and then the core checks to see what the currently selected action is, if any?

Yes. The idea is to simply let the user define his Items or NPCs by defining their appearance, "hot zones" (areas) and attach .esc scripts to them, and let the Core manage the events using the item's parameters. I usually try to determine what I'd need to re-create certain scenes of games I played. 
For example, I'd like to be able to define an area that triggers a cut-scene 
- the first time the player enters it
- the first time he just enters in a room
What would I need to create such gameplay? As a game designer, for each case, I'd either:
- define an Area2D on the ground, and attach an .esc script to it. The core tests if the player walks in it and runs the execution of the script if he does.
- attach an .esc script to the scene that tests if this is the first time the player enters it, and run the cut-scene if yes.

> Another thing this model reminded me of is that this would be a good opportunity to investigate the possibility of erasing some (all?) of the distinctions between items and and players and instead have either the core keep track of which item currently acts as the player or have that be a property of the item. The reason why this would be useful is because it would make it easier to do things like have an NPC walk to a specific point or even have multiple playable characters that you can switch between.

That's the kind of reflections I wanted ;) I definitely agree.
Basically, there is no difference between an NPC and an item. I'd personally set a property in the item (like `is_player`) so that it can be controlled by the core and thus, by script (the player 1 gets captured, switch to player 2).
NPC walk is a little bit different: the designer has to define waypoints and either use GDScript to make him move from one to the next one. I'd rather allow this in .esc scripts, or make it automatic if the NPC has a "waypoints" child node containing Point2D nodes.

**Comment by zero2000** [comment]
hello, watching on youtube I saw something interesting about godot 3.0 and maybe it could be implemented to be used in escoria [https://youtu.be/SD90vFweg-E](url)

**Comment by StraToN** [comment]
@zero2000 Thanks!
Are you talking about tweening the decoration scene elements like in the video?

**Comment by mjtorn** [comment]
The ideas here seem fine enough to someone who hasn't touched Escoria yet ;)

When I get around to testing Escoria, I'll implement a small Unity/AC prototype in it to compare what
the dev experience feels like. Something like making NPCs move would be nice to have made easy.

Supporting 3D is a goal I would postpone if it was my project, because rewriting is always 128 times harder than expected and trying to keep compatibility is even worse. Because 3D doesn't exist now, there are no problems with it.

What's the main setback here? Time, money, something else? What kind of time is estimated to be required to complete the rewrite?

**Comment by madeddy** [comment]
Hello folks,

i wanted to point you too some similar project i noticed recently: https://github.com/jeremi360/Ren (If you dont know it already) Not exactly escoria but maybe a little collaboration or something is possible.
Greets

**Comment by StraToN** [comment]
@mjtorn The main setback is indeed time. For information, money is never a setback since we're not a company, we only do this because we like to do it ;)
If the perimeter is well defined, which we're currently trying to achieve, the rewrite could be done in about 2 or 3 months, mostly for 2D. Some parts are common with 3D so they won't need different scripts in 2D and 3D.

**Comment by mjtorn** [comment]
@madeddy thanks for the tip, but unfortunately it doesn't look like what I need :( I'm sure you can collaborate on escoria, though :)

@StraToN I don't have any skin in the game or a horse in the race of rewriting, if the current master branch can let me at least prototype, but I've found that companies are good ways to enable doing things. Either that or some other way of covering expenses while working on something fun. Do you have plans in that direction to help with development?

**Comment by mjtorn** [comment]
I am flattered, but have to set the record straight, that I am very inexperienced in game development. My background is in completely different forms of programming than games, but making adventure games is a childhood dream and now I have a quasi-realistic shot at it. I studied the methodologies and such in my spare time, bought a handful of Unity assets and so on.

I'm not particularly _passionate_ about open-source, but I am most productive under Linux and there's a ton of other good reasons to use it beside personal gain. The problem with open-source ever so often is lack of commercial support, but I'm not super-convinced the support on the Unity side would be all that great for the assets. In the case of Escoria, I would gladly pay the price of Unity assets for similar functionality if that were possible.

There are quite a few downsides to Unity... Roughly everything's very much in binary format, making Git awkward, though maybe their cloud is a substitute. Then the later versions on Linux don't import my prototype properly, so I'm stuck with an older version unless I want to really delve into it. This is tough because I have my Dialogue System/Adventure Creator integration as work-in-progress, and I'd hate to restart everything just for that. Unity's also prone to crash on Linux, at least Debian Stable, which probably isn't even supported.

Then there's the part about black-box assets that are hard to integrate and/or should be implemented by myself.

So even if the Adventure Creator grass may be greener on the other side of the Unity fence, the fence might just be electrified and barb-wired, and I'm not super-convinced about the entire lawn.

However, if things take off and my only choice is to start cold anew with Unity, money starts talking and that's that with that.

And most definitely a rewrite should be in another branch! API stability is king, and a good way to ensure it is to never merge breaking changes into master without a properly announced release.

**Comment by madeddy** [comment]
@mjtorn The tip was more in general for the escoria rewrite. But no prob. :v:

@StraToN I am glad.

@fleskesvor My thoughts. The goals did appear similar to me, however do i not know if this ren'py style GUI and dev direction are compatible with your plans. You're informed and go talking. Mission accomplished.  

Another thought and question for you people:(crazy-idea-mode=on) 
Why this way for a adv game type support? Could it be possible to integrate it in some way in godot? I think in the direction of editor modes the user can switch on/off for specialized development. You get the template, toolset, editor config etc. for your (sub)type you need at the moment. No need to use the full godot-set and mainstream config if you do for example a very basic VN.
(Understandable description?)

Greets

**Comment by mjtorn** [comment]
On Fri, Feb 09, 2018 at 02:11:53PM -0800, madeddy wrote:
>
>Another thought and question for you people:(crazy-idea-mode=on) 
>Why this way for a adv game type support? Could it be possible to integrate it in some way in godot? I think in the direction of editor modes the user can switch on/off for specialized development. You get the template, toolset, editor config etc. for your (sub)type you need at the moment. No need to use the full godot-set and mainstream config if you do for example a very basic VN.
>(Understandable description?)

That means the engine would have a layer that has nothing to do
with the engine, and you'd fork the entire ordeal for development
of the adventure game part.

Someone will want a different thing and then there'd be another
framework either way, so even that's not a big win.

Of course if you want to ask the Godot developers, feel free to.

Getting Escoria stable is a good enough priority here right now IMO :)

**Comment by zero2000** [comment]
> @zero2000 Thanks!
Are you talking about tweening the decoration scene elements like in the video?

yes, It would be great to make them more attractive for those who wish to develop scum, they would see what they can do things that can not be done so easily in other engines , You should see Bladecoder Adventure Engine for maybe give a scum gui. it looks very good

**Comment by fleskesvor** [comment]
@mjtorn You have used Unity and Adventure Creator though, which gives you some insight into what should be expected of an adventure game framework. I've only ever dabbled a little with AGS, and that was many years ago.

@madeddy Thanks for the tip. 🙂 

I don't think Escoria should be integrated into Godot, but it's relatively easy to make editor plugins for Godot, which could be used to make Escoria a more integral part of the editor, if one wishes, something that has also been discussed. If that was to happen, I'm thinking that it would probably be a separate project that depends on Escoria, but not vice versa.

**Comment by bosinpai** [comment]
Hi,
I've been perusing the Escoria-in-Daïza manual, proofreading and upgrading to Godot 3 as I went.
I just realized that Escoria 3 might just not be ready yet.

Can you clarify its current status in the README: Is it usable right now? Is it stable or subject to change? What should new users use right now (godot2+esc2 or godot3+esc3)?

(bonus question: is Escoria plain incompatible with the HTML5 template, or is there a way to make it work? ^^)

**Comment by fleskesvor** [comment]
Hi, @bosinpai. Those are good questions.

I think we might be a couple of weeks off from landing a stable version of the framework. We've made some significant project structure changes since 2.1, with the goal of making it easier to manage framework and game separately, and ultimately to be able to make Escoria a Godot editor plugin. That work isn't finished yet, but it shouldn't change much once #114 is done.

The 2.1 version has the advantage of being stable, and exports work on Linux, macOS and Windows, but I don't think html5 exports work, judging by flossmanualsfr/escoria#12. It lacks several features that were added in 3.0 too, and migrating a project from 2.1 to 3.0 will be difficult because of the API changes in both Godot and Escoria.

As mentioned 3.0 isn't stable at the moment, but it is the only version that is being actively maintained, and we try to fix major bugs quickly. I haven't looked into html5 in a while, but it was not working the last time I checked. I can make it a priority if it's important for your current project. I would also like to have that working myself in time for Adventure Jam.

TL;DR I'd recommend Godot+Escoria 3, but be aware of future changes.

**Comment by bosinpai** [comment]
Thanks for the fast answer!
In the book I precised the versions used, and linked to the html5 issue in its Export section.
I'll have a look again at Esc3 later then, meanwhile I can continue the (much needed) English proofreading of the book :)

**Comment by fleskesvor** [comment]
Thanks for the help with the tutorial, @bosinpai! :slightly_smiling_face: Were you able to update it directly or do your changes get added to a review queue of some sort?

**Comment by sjb8100** [comment]
@fleskesvor I was glad to see the AC reference. I have searched the repo and this thread and other issues for 3D but I am not able to discern the state of 3D support in Escoria. My best guess is that @StraToN has some parts of 3D working in a diff branch? Is there a basic 3D example in works? If so, perhaps I could expand it to a robot and laser example in 3D with navigation (old school james bond and such.) thanks for your time and the updates.

**Comment by mjtorn** [comment]
@sjb8100 the most important part of Escoria is the scripting language, or at least it's my favorite ;) So starting with that, the first thing that comes to mind when thinking in three dimensions is that camera controls assume `x` and `y` but not `z`. Likewise for everything else involving movement. Then there'd need to be a command for rotating the camera, which would be disabled (or report an error) in 2D.

You could expand the esc parser to accept keyword arguments. In fact, the language would be better if it used _only_ keyword arguments (IMO) because the parsing is currently messy as everything looks at a `params` list and tries to figure out what's going on. That way 2D games could simply omit passing the `z` argument and we'd have compatibility. Want to take a shot at it?

I can't speak for what @StraToN might have stashed away somewhere, but I'm pretty sure it would be easy-ish to adapt most of the object types for 3D if the scripting language was better on board with it.
